File: typescript/examples/03-retrieve.ts
/**
 * Example 03 — Case retrieve with did-you-mean handling
 * Language: TypeScript (Node.js)
 *
 * Ambiguous queries return HTTP 200 + status "did_you_mean" — not an error.
 */
import { LawDiverClient } from "../src/client.js";

const client = new LawDiverClient();
const query = process.argv[2] ?? "410 U.S. 113";

let hit = await client.retrieve({ query });

if (hit.status === "did_you_mean") {
  console.log("Ambiguous — candidates:");
  for (const c of hit.candidates) {
    console.log(`  [${c.caseId}] ${c.bluebookCitation ?? c.caseName} (confidence=${c.confidence})`);
  }
  // Demo policy: pick the first candidate. In production, ask a human.
  const chosen = hit.candidates[0];
  if (!chosen) {
    console.log("No candidates returned.");
    process.exit(0);
  }
  console.log(`\nResolving with caseId=${chosen.caseId}…`);
  hit = await client.retrieve({ query, caseId: chosen.caseId });
}

if (hit.status === "not_found") {
  console.log("No match.", hit.corpusCaveat ?? "");
  console.log(`requestId: ${hit.requestId}`);
  process.exit(0);
}

if (hit.status === "ok") {
  const c = hit.case;
  console.log("OK");
  console.log(`  caseId: ${c.caseId ?? "(see payload)"}`);
  console.log(`  name:   ${(c as { caseName?: string }).caseName ?? "(see payload)"}`);
  console.log(`  pdfUrl: ${c.pdfUrl}`);
  console.log(`  requestId: ${hit.requestId}`);
}

File: python/examples/03_retrieve.py
"""Example 03 — Case retrieve with did-you-mean (Python)."""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from lawdiver import LawDiverClient

query = sys.argv[1] if len(sys.argv) > 1 else "410 U.S. 113"

with LawDiverClient() as client:
    hit = client.retrieve(query=query)

    if hit["status"] == "did_you_mean":
        print("Ambiguous — candidates:")
        for c in hit.get("candidates") or []:
            print(
                f"  [{c.get('caseId')}] {c.get('bluebookCitation') or c.get('caseName')} "
                f"(confidence={c.get('confidence')})"
            )
        chosen = (hit.get("candidates") or [None])[0]
        if not chosen:
            print("No candidates returned.")
            raise SystemExit(0)
        print(f"\nResolving with caseId={chosen['caseId']}…")
        hit = client.retrieve(query=query, case_id=chosen["caseId"])

if hit["status"] == "not_found":
    print("No match.", hit.get("corpusCaveat") or "")
    print(f"requestId: {hit['requestId']}")
elif hit["status"] == "ok":
    case = hit["case"]
    print("OK")
    print(f"  caseId: {case.get('caseId')}")
    print(f"  name:   {case.get('caseName')}")
    print(f"  pdfUrl: {case.get('pdfUrl')}")
    print(f"  requestId: {hit['requestId']}")
